File: spring/src/test/java/cardsproject/controller/cards/CardRulingControllerTest.java
package cardsproject.controller.cards;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.transaction.annotation.Transactional;

import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

@SpringBootTest
@AutoConfigureMockMvc
@Transactional
public class CardRulingControllerTest {

    @Autowired
    private MockMvc mockMvc;

    @Test
    void list_returns200() throws Exception {
        mockMvc.perform(get("/api/card_rulings"))
            .andExpect(status().isOk());
    }

    @Test
    void create_returns201() throws Exception {
        mockMvc.perform(post("/api/card_rulings")
            .contentType(MediaType.APPLICATION_JSON)
            .content("{ \"rulingText\": \"test\", \"publishedAt\": \"2024-01-01\", \"source\": \"test\" }"))
            .andExpect(status().isCreated());
    }

    @Test
    void show_returns200or404() throws Exception {
        mockMvc.perform(get("/api/card_rulings/1"))
            .andExpect(result -> {
                int status = result.getResponse().getStatus();
                assert status == 200 || status == 404;
            });
    }

    @Test
    void delete_returns204or404() throws Exception {
        mockMvc.perform(delete("/api/card_rulings/1"))
            .andExpect(result -> {
                int status = result.getResponse().getStatus();
                assert status == 204 || status == 404;
            });
    }
}

File: phoenix/test/cards/card_ruling_controller_test.exs
defmodule CardsProjectWeb.Cards.CardRulingControllerTest do
  use ExUnit.Case, async: false
  use CardsProjectWeb.ConnCase

  @valid_params %{
    "ruling_text" => "test",
    "published_at" => ~D[2024-01-01],
    "source" => "test"
  }

  describe "GET /api/card_rulings" do
    test "returns 200 with list", %{conn: conn} do
      conn = get(conn, "/api/card_rulings")
      assert json_response(conn, 200) |> is_list()
    end
  end

  describe "POST /api/card_rulings" do
    test "creates record and returns 201", %{conn: conn} do
      conn = post(conn, "/api/card_rulings", @valid_params)
      assert %{"id" => _id} = json_response(conn, 201)
    end
  end

  describe "GET /api/card_rulings/:id" do
    test "returns 200 for existing record", %{conn: conn} do
      {:ok, record} = CardsProject.Cards.create_card_ruling(@valid_params)
      conn = get(conn, "/api/card_rulings/#{record.id}")
      assert json_response(conn, 200)
    end
  end

  describe "PUT /api/card_rulings/:id" do
    test "updates and returns 200", %{conn: conn} do
      {:ok, record} = CardsProject.Cards.create_card_ruling(@valid_params)
      conn = put(conn, "/api/card_rulings/#{record.id}", @valid_params)
      assert json_response(conn, 200)
    end
  end

  describe "DELETE /api/card_rulings/:id" do
    test "deletes and returns 204", %{conn: conn} do
      {:ok, record} = CardsProject.Cards.create_card_ruling(@valid_params)
      conn = delete(conn, "/api/card_rulings/#{record.id}")
      assert response(conn, 204)
    end
  end
end
